src/BuildScriptGenerator/DotNetCore/GlobalJsonSdkResolver.cs

```diff
@@ -0,0 +1,371 @@
+﻿// --------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+// --------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+using Microsoft.Oryx.BuildScriptGenerator.Exceptions;
+using Microsoft.Oryx.Common.Extensions;
+
+namespace Microsoft.Oryx.BuildScriptGenerator.DotNetCore
+{
+    public class GlobalJsonSdkResolver
+    {
+        public const string LatestPatch = "latestPatch";
+        public const string Patch = "patch";
+        public const string LatestFeature = "latestFeature";
+        public const string Feature = "feature";
+        public const string LatestMinor = "latestMinor";
+        public const string Minor = "minor";
+        public const string LatestMajor = "latestMajor";
+        public const string Major = "major";
+        public const string Disable = "disable";
+
+        private readonly ILogger<GlobalJsonSdkResolver> _logger;
+
+        public GlobalJsonSdkResolver(ILogger<GlobalJsonSdkResolver> logger)
+        {
+            _logger = logger;
+        }
+
+        public string GetSatisfyingSdkVersion(GlobalJsonModel globalJson, IEnumerable<string> availableSdks)
+        {
+            /*
+                From spec:
+                If no global.json file is found, or global.json doesn't specify an SDK version nor an
+                allowPrerelease value, the highest installed SDK version is used (equivalent to setting rollForward
+                to latestMajor).
+            */
+            if (globalJson?.Sdk == null)
+            {
+                globalJson = new GlobalJsonModel();
+                globalJson.Sdk = new Sdk
+                {
+                    Version = "0.0.000",
+                    RollForward = LatestMajor,
```

Comment:
As per the spec, if there is no global.json file, then it should be considered as though there is one and that the roll forward policy is set to latest major, and the spec also says that if no setting is provided by the user for `allowPrerelease` flag, then it is assumed to be `true` by default, so that means prerelease versions must also be considered. 

This is going to be a problem for *our* case because as an end user you might have developed your 2.1/3.1 application locally using a 2.1/3.1 specific non-preview version sdks and it is also very common for people to *not* have global.json files. So based on the global.json spec, we should be considering latest major(including previews) for this app, which means we might try to build a 2.1/3.1 app using a `5.0.100-preview` sdk if that is the most available version in our storage. This might break existing users.

We could follow the following approaches:
1. Ask users to provide a global.json file
2. If there is no global.json file provided by the user, we will try getting the most recent *non-preview* version for that major version. i.e if the runtime version is `3.1`, then we will try getting an sdk version whose major is `3` but rest of the parts like minor, feature and patch would be the latest available one.

I prefer option `2.` above.

Reply:
> can dotnet core releases have breaking changes in their minor version?

No, .NET Core team would not do breaking changes in minor versions intentionally. Imagine this, currently if there is *no* global.json file, .NET Core cli picks up the *latest major* version, so what we are suggesting here we do is far safer than the default.